[CIR] Upstream global variable linkage types

mmha opened the pull request:
This change implements variable linkage types in ClangIR except for common linkage which requires Comdat support.

erichkeane commented on clang/include/clang/CIR/Dialect/IR/CIROpsEnums.h:
```diff
@@ -0,0 +1,119 @@
+//===- CIROpsEnumsDialect.h - MLIR Dialect for CIR ----------------------*- C++
```
This comment doesn't match the file name.

bcardosolopes replied:
TIL

erichkeane commented on clang/include/clang/CIR/Dialect/IR/CIROpsEnums.h:
```diff
@@ -0,0 +1,119 @@
+//
+//===----------------------------------------------------------------------===//
+//
+// This file declares the Target dialect for CIR in MLIR.
```
Does it?  Seems this should better match the file.

erichkeane commented on clang/include/clang/CIR/Interfaces/CIROpInterfaces.h:
```diff
@@ -0,0 +1,29 @@
+#include "clang/AST/Mangle.h"
+#include "clang/CIR/Dialect/IR/CIROpsEnums.h"
+
+namespace cir {} // namespace cir
```
Can we remove these empty namespace cirs?  We can add them if we need them, but they seem sillly at the moment.

erichkeane reviewed (commented):
Only nits from me.  I'll add other reviewers however.

andykaylor commented on clang/include/clang/CIR/Interfaces/CIROpInterfaces.h:
```diff
@@ -0,0 +1,29 @@
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef MLIR_INTERFACES_CIR_OP_H_
```
```suggestion
#ifndef CLANG_CIR_INTERFACES_CIR_OP_H
```

andykaylor commented on clang/include/clang/CIR/Interfaces/CIROpInterfaces.td:
```diff
@@ -0,0 +1,63 @@
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef MLIR_CIR_OP_INTERFACES
```
```suggestion
#ifndef CLANG_CIR_INTERFACES_CIROPINTERFACES_TD
```

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenModule.cpp:
```diff
@@ -176,6 +177,18 @@ void CIRGenModule::emitGlobalVarDefinition(const clang::VarDecl *vd,
+    // Set CIR linkage and DLL storage class.
+    varOp.setLinkage(linkage);
+
+    if (linkage == cir::GlobalLinkageKind::CommonLinkage) {
```
No need for curly braces here. They're omitted for single-line scopes in LLVM's coding style.

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenModule.cpp:
```diff
@@ -210,6 +223,193 @@ void CIRGenModule::emitGlobalDefinition(clang::GlobalDecl gd,
+  llvm_unreachable("No such linkage");
+}
+
+// TODO(CIR): this could be a common method between LLVM codegen.
```
Can you do that as part of this change? Maybe move this into ASTContext?

andykaylor commented on clang/include/clang/CIR/MissingFeatures.h:
```diff
@@ -36,6 +36,10 @@ struct MissingFeatures {
   static bool opGlobalConstant() { return false; }
   static bool opGlobalAlignment() { return false; }
   static bool opGlobalLinkage() { return false; }
```
Can this be removed now?

andykaylor commented on clang/lib/CIR/Interfaces/CIROpInterfaces.cpp:
```diff
@@ -0,0 +1,22 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
```
I notice we're missing file descriptions in header blocks of several of the files you're adding here. I don't know how much people care about that, but the coding standards say there say there should be one. Many existing files that have it don't follow the doxygen format described in the coding standards.

https://llvm.org/docs/CodingStandards.html#file-headers

andykaylor reviewed (commented):
Can you also update `clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp` to use the linkage? There are two places marked by `cir::MissingFeatures::opGlobalLinkage()` that would need to be updated, plus (hopefully) test changes.

It would also be good to have tests for more linkage types. I'm not sure there is a single existing test that verifies these. I found some in `clang/test/CIR/CodeGen/weak.c` in the incubator and some other examples in `clang/test/CodeGen/global-decls.c`. Perhaps a new `global-linkage.cpp` test could be added that cover all the variations we currently support?

bcardosolopes reviewed (commented):
Great to see this PR Morris! You should also add CIR <-> CIR tests for the linkage types, incubator has some of them you can copy n paste.

bcardosolopes commented on clang/test/CIR/global_var_linkage.cpp:
```diff
@@ -0,0 +1,11 @@
+// Linkage types of global variables
+// RUN: %clang_cc1 -std=c++20 -triple x86_64-unknown-linux-gnu -fclangir -emit-cir %s -o -  | FileCheck %s
+
+int aaaa;
```
This new tests uses_this_convention and while the existing-uses-this, I suggest you-keep-this-uniform.

dkolsen-pgi commented on clang/test/CIR/global-var-simple.cpp:
```diff
@@ -2,100 +2,100 @@
+// CHECK: cir.global external @ub48 : !cir.int<u, 48>
 
 bool boolfalse = false;
 // CHECK: cir.global @boolfalse = #false
```
Doesn't this one also need "external" added to the expected CIR?

dkolsen-pgi reviewed (commented):
Most of the things I would have commented were already mentioned by someone else.  Just a couple small additional items.

xlauko commented on clang/include/clang/CIR/Interfaces/CIROpInterfaces.td:
```diff
@@ -0,0 +1,63 @@
+    let methods = [
+      InterfaceMethod<"",
+      "bool", "hasAvailableExternallyLinkage", (ins), [{}],
+      /*defaultImplementation=*/[{ return false; }]
```
I fixed these in incubator: https://github.com/llvm/clangir/pull/1423 and refined some additional class declarations. The changes from that PR can be applied here as well.

mmha replied:
I applied your changes in my patch except for the additions to `CallOp` which doesn't exist, yet.

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenModule.cpp:
```diff
@@ -210,6 +223,193 @@ void CIRGenModule::emitGlobalDefinition(clang::GlobalDecl gd,
+}
+
+// TODO(CIR): this could be a common method between LLVM codegen.
+static bool isVarDeclStrongDefinition(const ASTContext &astContext,
```
That sounds like a good idea.  IMO we should have a directory of `common` codegen stuff like this that we can use to extract stuff from llvm-ir codegen and use from CIR.  This seems like a good candidate to start that.

bcardosolopes replied:
I have annotated tons of these functions around, I think they resemble the idea of `ASTCodegenHelpers`, since they allow AST queries that have no dependency to codegen and are also useful to other tools (each these days implement their own). Not sure where exactly the implementation should live, but it would be nice if those can be exposed on a public header

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenModule.cpp:
```diff
@@ -210,6 +223,193 @@ void CIRGenModule::emitGlobalDefinition(clang::GlobalDecl gd,
+  return false;
+}
+
+cir::GlobalLinkageKind CIRGenModule::getCIRLinkageForDeclarator(
```
I answered above, but if modifying current CodeGen mildly is necessary in order to make the CIR process easier (and to improve the amount of shared code), then I'm 100% for it.

bcardosolopes replied:
> this would require changes to the incubator and CodeGen first, as some functions are private methods or rely on langOpts from CodeGen. So this is more of a question for @erichkeane and @bcardosolopes—would this be feasible?

I'm not a big fan of depending on LLVM stuff at the CIR level, given we can get all info we need from the AST and even though it'd be a practical solution - my take is to keep that at the lowering level.

> This is likely a general policy question: how tightly should CodeGens be coupled, and should we modify main CodeGen to better support API reuse for CIRCodeGen? I expect many similar copies, independent of the actual CIR and MLIR, to appear during upstreaming.

I believe reuse is always great, but IMO I don't think we need to generate extra work for upstreaming on anything other than things like a `ASTCodegenHelpers` atm.

xlauko committed:
cherry-pick 0bedc28 Clean up and complete CIRGlobalValueInterface methods

mmha committed:
Address more review comments

- Add header comments to CIROpInterfaces
- Fix test clang-cir-ir-global

erichkeane reviewed (approved):
I'm happy when everyone else is, the refactor to extract common functions is valuable but can be/should be done in a separate patch.

bcardosolopes reviewed (approved):
LGTM too

mmha committed:
Add CIR -> CIR test for global variable linkage

xlauko commented on clang/include/clang/CIR/Interfaces/CIROpInterfaces.td:
```diff
@@ -0,0 +1,103 @@
+         InterfaceMethod<"", "bool", "isWeakForLinker", (ins), [{}],
+                         /*defaultImplementation=*/[{
+        return cir::isWeakForLinker($_op.getLinkage());
+      }]>];
```
Is this intended indentation?

mmha replied:
So to speak, yes. I reformatted the file with clang-format.

dkolsen-pgi replied:
Maybe that's why clang-format isn't commonly used on table-gen files.  That indentation is wrong and makes it (a tiny bit) harder to read.

mmha committed:
Reformat CIROpInterfaces.td

xlauko commented on clang/include/clang/CIR/Dialect/IR/CIROpsEnums.h:
```diff
@@ -0,0 +1,119 @@
+//===----------------------------------------------------------------------===//
+//-*-===//
```
```suggestion
```

mmha committed:
Update clang/include/clang/CIR/Dialect/IR/CIROpsEnums.h

Co-authored-by: Henrich Lauko <xlauko@mail.muni.cz>

mmha committed:
[CIR] Upstream global variable linkage types

This change implements variable linkage types in ClangIR except for common linkage which requires Comdat support.

mmha committed:
Addressed some review feedback

- Various formatting changes
- Changed file header comments
- Changed include guards
- Added global linkage to LLVM lowering

bcardosolopes commented:
Looks like ready to land, @mmha do you have landing permission?

mmha commented:
Great! I don't have committer status, yet, if that's your question.
